Packages/ClientRuntime/Sources/Waiters/Matcher.swift

```diff
@@ -0,0 +1,45 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+extension Acceptor {
+
+    /// `Acceptor.Matcher` is a Swift-native equivalent of Smithy matchers:
+    /// https://awslabs.github.io/smithy/2.0/additional-specs/waiters.html#matcher-union
+    public enum Matcher {
+        case output((Output) -> Bool)
+        case inputOutput((Input, Output) -> Bool)
+        case success(Bool)
+        case errorType((Error) -> Bool)
+
+        func isAMatch(input: Input, output: Output?, error: Error?) -> Bool {
```

Comment:
Is it acceptable that this implementation allows for more flexibility than the [spec defines](https://awslabs.github.io/smithy/2.0/additional-specs/waiters.html#matcher-union)
For example, this implementation allows for the caller to define "has a match" as any logic they want, where as the spec has tight controls and restricts it to a set number (currently 4) comparators. etc.

Reply:
Unless someone really goes deep-diving, matchers should always be code-generated, so I'm not too worried about a friendly interface at this point